### PR TITLE
Soft 376 i2c write python

### PR DESCRIPTION
**Pull request by abdullahs26**

### CODE REVIEW

File: projects/baby_driver/scripts/i2c_write.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    Args:
+        port: port of the I2C to write to (1 or 2)
+        address: I2C address to write to (0-255)
+        tx_bytes: list of bytes to write over I2C (a list of ints 0-255)
+        reg: the register to write to, or None if no register is required
+    Raises:
+        Exception: if a non-zero status code was received
+    """
+    if port not in (1,0):
+        raise ValueError("Expected port of 0 (I2C_PORT_1) or 1 (I2C_PORT_2)")
```
Comment (copperium) [inline projects/baby_driver/scripts/i2c_write.py:18]: It's a bit confusing to make users enter 0 for port 1 or 1 for port 2, eh? Better I think would be to take 1 or 2 for port 1 or port 2 respectively, then subtract 1 before packing the data.

File: projects/baby_driver/scripts/i2c_write.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    if address <0 or address > 255:
+        raise ValueError("Expected address between 0 and 255")
+    # pylint: disable=consider-using-enumerate
+    for i in range(len(tx_bytes)):
+        if tx_bytes[i] <0 or tx_bytes[i] > 255:
+            raise ValueError("Expected list of bytes between 0 and 255")
```
Comment (copperium) [inline projects/baby_driver/scripts/i2c_write.py:24]: Ah, I thought pylint was talking about your `range(0, len(tx_bytes), 7)` construction below with its consider-using-enumerate warning. It would be more idiomatic ("pythonic") to iterate directly over the list with `for byte in tx_bytes`, since you're not using the index beyond indexing into the list, but I think this might be the most pythonic way to do this condition:
```py
if not all(0 <= byte < 256 for byte in tx_bytes):
  raise ValueError("Expected list of bytes between 0 and 255")
```
This works! The thing in the brackets is a [generator comprehension](https://www.pythonlikeyoumeanit.com/Module2_EssentialsOfPython/Generators_and_Comprehensions.html) (like a list comprehension if you've seen those, but without brackets and not computing elements until they're asked for), `all` is a function which returns True if and only if all the values of its argument are True, and `0 <= byte < 256` is equivalent to `0 <= byte and byte < 256` because Python is magic.

This way you can remove the `# pylint: disable` comment.

File: projects/baby_driver/scripts/test_i2c_write.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+            self.assertEqual(None, self.channel)
+            self.assertEqual(BABYDRIVER_CAN_MESSAGE_ID, self.msg_id)
+            self.assertEqual(BABYDRIVER_DEVICE_ID, self.device_id)
+
+            i2c_write(1, 10, [7, 5, 12], 255)
+            self.assertEqual(9, self.babydriver_id)
+            self.assertEqual([7, 5, 12], self.data)
+            self.assertEqual(None, self.channel)
+            self.assertEqual(BABYDRIVER_CAN_MESSAGE_ID, self.msg_id)
+            self.assertEqual(BABYDRIVER_DEVICE_ID, self.device_id)
```
Comment (copperium) [inline projects/baby_driver/scripts/test_i2c_write.py:146]: Because of the indentation here, all of these asserts are actually inside the `test_parameter` function, so they aren't run. You'll need to deindent them so they run.

File: projects/baby_driver/scripts/test_i2c_write.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+        self.assertEqual(9, self.babydriver_id)
+        self.assertEqual([7, 5, 12, 14, 15, 18, 9], self.data)
+        self.assertEqual(None, self.channel)
+        self.assertEqual(BABYDRIVER_CAN_MESSAGE_ID, self.msg_id)
+        self.assertEqual(BABYDRIVER_DEVICE_ID, self.device_id)
+
+    @patch('can_util.send_message')
+    @patch('can_util.next_message')
+    def test_send_message_max(self, mock_next_message, mock_send_message):
+        """Tests accuracy of minimum parameters passed into can_util.send_message"""
```
Comment (copperium) [inline projects/baby_driver/scripts/test_i2c_write.py:94]: I think you forgot to change minimum -> maximum in this docstring.

File: projects/baby_driver/scripts/test_i2c_write.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+        self.assertRaises(ValueError,i2c_write, 1, -1, [0], 1)
+        self.assertRaises(ValueError,i2c_write, 1, 0, [266], 1)
+        self.assertRaises(ValueError,i2c_write, 1, 0, [0], 266)
+        self.assertRaises(ValueError,i2c_write, 2, 0, [0], -1)
+
+        # Tests fail conditions when there is not a registered write, (reg=None)
+        self.assertRaises(ValueError,i2c_write, 2, 0, [0])
+        self.assertRaises(ValueError,i2c_write, 1, 266, [0])
+        self.assertRaises(ValueError,i2c_write, 1, -1, [0])
+        self.assertRaises(ValueError,i2c_write, 1, 0, [266])
```
Comment (copperium) [inline projects/baby_driver/scripts/test_i2c_write.py:165]: You should use 256 instead of 266 if you want to test the exact fail condition boundary

File: projects/baby_driver/scripts/test_i2c_write.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+            self.assertEqual(None, self.channel)
+            self.assertEqual(BABYDRIVER_CAN_MESSAGE_ID, self.msg_id)
+            self.assertEqual(BABYDRIVER_DEVICE_ID, self.device_id)
+
+            i2c_write(1, 255, [7, 5, 12, 7, 5, 12, 7, 255, 255, 255, 255, 255, 255, 255], 1)
+            self.assertEqual(9, self.babydriver_id)
+            self.assertEqual([255, 255, 255, 255, 255, 255, 255], self.data)
+            self.assertEqual(None, self.channel)
+            self.assertEqual(BABYDRIVER_CAN_MESSAGE_ID, self.msg_id)
+            self.assertEqual(BABYDRIVER_DEVICE_ID, self.device_id)
```
Comment (copperium) [inline projects/baby_driver/scripts/test_i2c_write.py:132]: You should also check that the first bytes of the data are as you expect - either append to `self.data` in `test_parameter` and set it to `[]` before every `i2c_write` call, or at least keep a count of how many times `test_parameter` is called and check it.

File: projects/baby_driver/scripts/test_i2c_write.py [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+            self.channel = channel
+            self.msg_id = msg_id
+            self.device_id = device_id
+
+            mock_send_message.side_effect = test_parameter
+            mock_next_message.return_value.data = [0,0]
+
+            # Tests high parameters for can_util.send_message
+            i2c_write(1, 10, [7, 5, 12, 7, 5, 12, 8, 1, 2, 3, 4, 5, 6, 7], 1)
+            self.assertEqual(9, self.babydriver_id)
```
Comment (copperium) [inline projects/baby_driver/scripts/test_i2c_write.py:121]: Import `BabydriverMessageId` from `message_defs` and use `BabydriverMessageId.I2C_WRITE_DATA` instead of 9 here and elsewhere.

File: projects/baby_driver/scripts/test_i2c_write.py [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+        # Tests maximum values for port, address, tx_bytes and reg
+        i2c_write(1,1,[7, 5, 12],1)
+        i2c_write(0,255,[7, 5, 12],1)
+        i2c_write(0,1,[255, 255, 255],1)
+        i2c_write(0,1,[7, 5, 12],255)
+
+    @patch('can_util.send_message')
+    @patch('can_util.next_message')
+
+    def test_send_message_min(self, mock_next_message, mock_send_message):
```
Comment (copperium) [inline projects/baby_driver/scripts/test_i2c_write.py:36]: You never test here or in `test_send_message_min` that the command message is as you expect, you should figure out a way to test that.

File: projects/baby_driver/scripts/test_i2c_write.py [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+        i2c_write(1, 10, [7, 5, 12, 14, 15, 18, 9], 1)
+        self.assertEqual(9, self.babydriver_id)
+        self.assertEqual([7, 5, 12, 14, 15, 18, 9], self.data)
+        self.assertEqual(None, self.channel)
+        self.assertEqual(BABYDRIVER_CAN_MESSAGE_ID, self.msg_id)
+        self.assertEqual(BABYDRIVER_DEVICE_ID, self.device_id)
+
+    @patch('can_util.send_message')
+    @patch('can_util.next_message')
+    def test_send_message_max(self, mock_next_message, mock_send_message):
```
Comment (copperium) [inline projects/baby_driver/scripts/test_i2c_write.py:93]: A couple edge cases I'd like to see tests for:
* `len(tx_bytes) == 255`
* in `test_fail_conditions`: failing when `len(tx_bytes) == 256`
* `len(tx_bytes)` is greater than 7 but not divisible by 7

File: projects/baby_driver/scripts/test_i2c_write.py [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+
+#pylint: disable=unused-argument
+class TestI2CWrite(unittest.TestCase):
+    """Test i2c_write function"""
+
+    @patch('can_util.send_message')
+    @patch('can_util.next_message')
+    #pylint: disable =no-self-use
+
+    def test_i2c_write_parameters(self, mock_next_messaage, mock_send_message):
```
Comment (copperium) [inline projects/baby_driver/scripts/test_i2c_write.py:16]: Typo: `mock_next_messaage`

**Review by copperium — CHANGES_REQUESTED** [review]
Nice, a few things to fix, mostly in the tests.

We don't have a Python autoformatter yet (coming soon(TM)) but the [Google Python style guide](https://google.github.io/styleguide/pyguide.html) provides a good basis for code formatting. In particular:
* a space on either side of binary operators like `<`
* a space after commas
* trailing commas after the last parameter when splitting a function call/declaration onto multiple lines
* no blank line after decorators (with `@`)

I'm also not sure why every file has an indented line at the end, maybe your editor's doing it automatically? In any case, that should be fixed.

File: projects/baby_driver/scripts/i2c_write.py [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+    Args:
+        port: port of the I2C to write to (1 or 2)
+        address: I2C address to write to (0-255)
+        tx_bytes: list of bytes to write over I2C (a list of ints 0-255)
+        reg: the register to write to, or None if no register is required
+    Raises:
+        Exception: if a non-zero status code was received
+    """
+    if port not in (1,2):
+        raise ValueError("Expected port of 0 (I2C_PORT_1) or 1 (I2C_PORT_2)")
```
Comment (copperium) [inline projects/baby_driver/scripts/i2c_write.py:18]: Fix this comment to align with the 1 or 2 for the port

File: projects/baby_driver/scripts/test_i2c_write.py [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+        self.assertEqual(BabydriverMessageId.I2C_WRITE_DATA, self.babydriver_id)
+        self.assertEqual([0, 10, 8, 1, 255], self.command)
+        self.assertEqual([7, 5, 12, 7, 5, 12, 7, 8], self.data)
+        self.assertEqual(None, self.channel)
+        self.assertEqual(BABYDRIVER_CAN_MESSAGE_ID, self.msg_id)
+        self.assertEqual(BABYDRIVER_DEVICE_ID, self.device_id)
+
+        # Tests for len(tx_bytes) == 255
+        self.data = []
+        testlist = [1] * 175
```
Comment (copperium) [inline projects/baby_driver/scripts/test_i2c_write.py:225]: This is 175? Not 255?

File: projects/baby_driver/scripts/i2c_write.py [hunk 13]
```diff
@@ -0,0 +1,49 @@
+"""Python implementation of the i2c_write function"""
+import can_util
+from message_defs import BabydriverMessageId
+
+
+def i2c_write(port,address, tx_bytes, reg=None):
```
Comment (copperium) [inline projects/baby_driver/scripts/i2c_write.py:6]: Style point: please put a space after the comma after `port,`

File: projects/baby_driver/scripts/i2c_write.py [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+        port: port of the I2C to write to (1 or 2)
+        address: I2C address to write to (0-255)
+        tx_bytes: list of bytes to write over I2C (a list of ints 0-255)
+        reg: the register to write to, or None if no register is required
+    Raises:
+        Exception: if a non-zero status code was received
+    """
+    if port not in (1,2):
+        raise ValueError("Expected port of 0 (I2C_PORT_1) or 1 (I2C_PORT_2)")
+    if address <0 or address > 255:
```
Comment (copperium) [inline projects/baby_driver/scripts/i2c_write.py:19]: Style nit: space after `<` please

File: projects/baby_driver/scripts/i2c_write.py [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+        raise ValueError("Expected address between 0 and 255")
+    if not all(0 <= byte < 256 for byte in tx_bytes):
+        raise ValueError("Expected list of bytes between 0 and 255")
+
+    is_reg = 0
+    if reg is None:
+        reg = 0
+    else:
+        is_reg = 1
+    if reg <0 or reg > 255:
```
Comment (copperium) [inline projects/baby_driver/scripts/i2c_write.py:29]: Same, space after `<` please

File: projects/baby_driver/scripts/i2c_write.py [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+        raise ValueError("Expected register to write to between 0 and 255")
+
+    can_pack = can_util.can_pack([(port-1, 1), (address, 1),
+    (len(tx_bytes), 1), (is_reg, 1), (reg, 1)])
```
Comment (copperium) [inline projects/baby_driver/scripts/i2c_write.py:33]: Style nit: this looks at first glance like two unrelated lines of code, please indent the second line:
```py
can_pack = can_util.can_pack([(port-1, 1), (address, 1),
    (len(tx_bytes), 1), (is_reg, 1), (reg, 1)])
```

File: projects/baby_driver/scripts/i2c_write.py [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+
+    can_util.send_message(
+        babydriver_id=BabydriverMessageId.I2C_WRITE_COMMAND,
+        data=can_pack
```
Comment (copperium) [inline projects/baby_driver/scripts/i2c_write.py:37]: Style nit: I'd prefer a comma at the end of this line

File: projects/baby_driver/scripts/i2c_write.py [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+    )
+
+    # Sends CAN message
+    for i in range(0, len(tx_bytes), 7):
+        can_util.send_message(
+            babydriver_id=BabydriverMessageId.I2C_WRITE_DATA,
+            data=tx_bytes[i:i+7]
+            )
```
Comment (copperium) [inline projects/baby_driver/scripts/i2c_write.py:45]: Style nit: the closing parenthesis should be on the same level of indentation as the `can_util.send_message`, and there should be a comma at the end of the `data=` line.

File: projects/baby_driver/scripts/message_defs.py [hunk 19]
```diff
@@ -11,7 +11,6 @@
 class BabydriverMessageId:
     """
     An enumeration of babydriver IDs, which go in the first uint8 in a babydriver CAN message.
-
```
Comment (copperium) [inline projects/baby_driver/scripts/message_defs.py:14]: Not sure why you deleted this?

File: projects/baby_driver/scripts/message_defs.py [hunk 20]
```diff
@@ -22,3 +21,6 @@ class BabydriverMessageId:
     GPIO_GET_DATA = 3
     ADC_READ_COMMAND = 4
     ADC_READ_DATA = 5
+    I2C_WRITE_COMMAND = 8
+    I2C_WRITE_DATA = 9
+
```
Comment (copperium) [inline projects/baby_driver/scripts/message_defs.py:26]: Not sure why you added a blank line with some indentation at the end?

File: projects/baby_driver/scripts/repl_setup.py [hunk 21]
```diff
@@ -2,8 +2,8 @@
 This script is run just before the Python REPL opens when you run `make babydriver`.
 It imports everything that's visible to the REPL by default and sets up the default CAN channel.
 """
-
```
Comment (copperium) [inline projects/baby_driver/scripts/repl_setup.py:5]: Same, not sure why you deleted this or why there's a blank line with indentation added at the end of the file

File: projects/baby_driver/scripts/test_i2c_write.py [hunk 22]
```diff
@@ -0,0 +1,264 @@
...
+from message_defs import BabydriverMessageId
+
+#pylint: disable=unused-argument
+class TestI2CWrite(unittest.TestCase):
+    """Test i2c_write function"""
+
+    @patch('can_util.send_message')
+    @patch('can_util.next_message')
+    #pylint: disable =no-self-use
+
```
Comment (copperium) [inline projects/baby_driver/scripts/test_i2c_write.py:16]: Style nit: there shouldn't be a blank line between the decorators (with `@`) and the function declaration, also no space before the `=` sign in this comment, and a space after the `#`:  `# pylint: disable=no-self-use`

File: projects/baby_driver/scripts/test_i2c_write.py [hunk 23]
```diff
@@ -0,0 +1,264 @@
...
+        # Tests maximum values for port, address, tx_bytes and reg
+        i2c_write(2,1,[7, 5, 12],1)
+        i2c_write(1,255,[7, 5, 12],1)
+        i2c_write(1,1,[255, 255, 255],1)
+        i2c_write(1,1,[7, 5, 12],255)
+
+    @patch('can_util.send_message')
+    @patch('can_util.next_message')
+
```
Comment (copperium) [inline projects/baby_driver/scripts/test_i2c_write.py:36]: Style nit: no blank line here

File: projects/baby_driver/scripts/test_i2c_write.py [hunk 24]
```diff
@@ -0,0 +1,264 @@
...
+        self.msg_id = None
+        self.device_id = None
+
+        def test_parameter(
+            babydriver_id=None,
+            data=None,
+            channel=None,
+            msg_id=BABYDRIVER_CAN_MESSAGE_ID,
+            device_id=BABYDRIVER_DEVICE_ID,
+            ):
```
Comment (copperium) [inline projects/baby_driver/scripts/test_i2c_write.py:54]: Style nit: this `):` should be deindented one level

File: projects/baby_driver/scripts/test_i2c_write.py [hunk 25]
```diff
@@ -0,0 +1,264 @@
...
```
Comment (copperium) [inline projects/baby_driver/scripts/test_i2c_write.py:126]: Same, this `):` should be deindented one level

File: projects/baby_driver/scripts/test_i2c_write.py [hunk 26]
```diff
@@ -0,0 +1,264 @@
...
+        self.assertEqual(BabydriverMessageId.I2C_WRITE_DATA, self.babydriver_id)
+        self.assertEqual([0, 10, 3, 1, 255], self.command)
+        self.assertEqual([7, 5, 12], self.data)
+        self.assertEqual(None, self.channel)
+        self.assertEqual(BABYDRIVER_CAN_MESSAGE_ID, self.msg_id)
+        self.assertEqual(BABYDRIVER_DEVICE_ID, self.device_id)
+
+    @patch('can_util.send_message')
+    @patch('can_util.next_message')
+
```
Comment (copperium) [inline projects/baby_driver/scripts/test_i2c_write.py:183]: Style nit: no blank line here

File: projects/baby_driver/scripts/test_i2c_write.py [hunk 27]
```diff
@@ -0,0 +1,264 @@
...
+        self.assertEqual(BabydriverMessageId.I2C_WRITE_DATA, self.babydriver_id)
+        self.assertEqual([0, 10, 175, 1, 1], self.command)
+        self.assertEqual(testlist, self.data)
+        self.assertEqual(None, self.channel)
+        self.assertEqual(BABYDRIVER_CAN_MESSAGE_ID, self.msg_id)
+        self.assertEqual(BABYDRIVER_DEVICE_ID, self.device_id)
+
+    @patch('can_util.send_message')
+    @patch('can_util.next_message')
+
```
Comment (copperium) [inline projects/baby_driver/scripts/test_i2c_write.py:236]: Style nit: no blank line here

**Review by copperium — CHANGES_REQUESTED** [review]
Sorry for the million comments, almost all of these are simple style nitpicks. I really need to set up a Python autoformatter...

**Review by copperium — APPROVED** [review]
